src/client/mod.rs

```diff
@@ -140,31 +164,54 @@ impl InfluxDbClient {
 
         let client = if let Some(_) = any_value.downcast_ref::<InfluxDbReadQuery>() {
             let read_query = query.get();
-            let encoded: String = form_urlencoded::Serializer::new(String::new())
-                .append_pair("db", self.database_name())
-                .append_pair("q", &read_query)
-                .finish();
-            let http_query_string = format!(
-                "{url}/query?{encoded}",
-                url = self.database_url(),
-                encoded = encoded
-            );
+   
+            let mut parameters = vec![
+                ("db", self.database_name()),
+                ("q", &read_query),
+            ];
+            
+            if let Some(auth) = self.auth() {
+                parameters.push(("u", auth.username.as_str()));
+                parameters.push(("p", auth.password.as_str()));
+            }
+
+            let url = match Url::parse_with_params(format!("{url}/query", url = self.database_url()).as_str(), parameters) {
+                Ok(url) => url,
+                Err(err) => {
+                    let error = InfluxDbError::InvalidQueryError {
+                        error: format!("{}", err),
+                    };
+                    return Box::new(future::err::<String, InfluxDbError>(error));
+                }
+            };
             if read_query.contains("SELECT") || read_query.contains("SHOW") {
-                Client::new().get(http_query_string.as_str())
+                Client::new().get(url)
             } else {
-                Client::new().post(http_query_string.as_str())
+                Client::new().post(url)
             }
         } else if let Some(write_query) = any_value.downcast_ref::<InfluxDbWriteQuery>() {
+            let precision_modfier = write_query.get_precision_modifier();
+            let mut parameters = vec![
+                ("db", self.database_name()),
+                ("precision", precision_modfier.as_str()),
+            ];
+            
+            if let Some(auth) = self.auth() {
+                parameters.push(("u", auth.username.as_str()));
+                parameters.push(("p", auth.password.as_str()));
+            } 
+
+            let url = match Url::parse_with_params(format!("{url}/write", url = self.database_url()).as_str(), parameters) {
+                Ok(url) => url,
+                Err(err) => {
+                    let error = InfluxDbError::InvalidQueryError {
+                        error: format!("{}", err),
```

Comment:
There should be a custom error type implemented here, or at least the message should be more descriptive, even when this error is unlikely to happen.

Reply:
What about?
```
#[fail(display = "failed to build URL: {}", error)]
/// Error happens when a query is invalid
UrlConstructionError { error: String },
```